### PR TITLE
[AMORO-1940] Support list partitions via flink catalog

### LINKED ISSUE #1940: [Improvement]: Support list partitions via flink catalog
### Search before asking

- [X] I have searched in the [issues](https://github.com/NetEase/amoro/issues?q=is%3Aissue) and found no similar issues.

### What would you like to be improved?

Now the flink catalog is not implemented `public List<CatalogPartitionSpec> listPartitions(ObjectPath tablePath)` interface; so `SHOW PARTITIONS` statement is not supported.

### How should we improve?

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Subtasks

_No response_

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by huyuanfeng2018**
## Why are the changes needed?

Close #1940

## Brief change log

- Implement the `listpartitions` method in `ArcticCatalog`

## How was this patch tested?

- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

- Does this pull request introduce a new feature? (no)
- If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by YesOrNo828** [comment]
@huyuanfeng2018  Thanks for your contribution. The CI failed due to the checkstyle. Could you take a look?

**Comment by huyuanfeng2018** [comment]
> @huyuanfeng2018 Thanks for your contribution. The CI failed due to the checkstyle. Could you take a look?

done.

File: flink/v1.12/flink/src/test/java/com/netease/arctic/flink/TestArcticCatalogTablePartitions.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    Assert.assertEquals("Should produce the expected catalog partition specs.", list, expected);
+  }
+
+
+  @Test
+  public void testListPartitionsKeyedTable() throws TableNotPartitionedException {
+    List<Object[]> data = new LinkedList<>();
+    data.add(new Object[]{1, "mark", "2023-10-01"});
+    data.add(new Object[]{2, "Gerry", "2023-10-02"});
+    data.add(new Object[]{RowKind.DELETE, 2, "Gerry", "2023-10-02"});
```
Comment (huyuanfeng2018) [inline flink/v1.12/flink/src/test/java/com/netease/arctic/flink/TestArcticCatalogTablePartitions.java:101]: Like this, first insert a piece of data with ID 2 and then delete it. ListPartition cannot determine whether the partition  `2023-10-02` really exists.
@YesOrNo828 
 WDYT?

Comment (YesOrNo828) [reply flink/v1.12/flink/src/test/java/com/netease/arctic/flink/TestArcticCatalogTablePartitions.java:101]: @huyuanfeng2018 I don't think there's an excellent way to return the data's partitions. Additionally, the issue is present in the iceberg v2 table.
Perhaps we can add a comment to the method to clarify that the partitions returned in this scenario encompass all the partitions related to writing the data, including -D and -U.

cc @zhoujinsong  WDYT?

Comment (zhoujinsong) [reply flink/v1.12/flink/src/test/java/com/netease/arctic/flink/TestArcticCatalogTablePartitions.java:101]: I think it's okay to return this partition in this situation.

File: flink/v1.12/flink/src/main/java/com/netease/arctic/flink/catalog/ArcticCatalog.java [hunk 2]
```diff
@@ -324,8 +334,40 @@ public void alterTable(ObjectPath tablePath, CatalogBaseTable newTable, boolean
...
+    List<Table> tables = new LinkedList<>();
+
+    if (arcticTable.isKeyedTable()) {
+      tables.add(arcticTable.asKeyedTable().changeTable());
+      tables.add(arcticTable.asKeyedTable().baseTable());
+    } else {
+      tables.add(arcticTable.asUnkeyedTable());
+    }
+    for (Table table : tables) {
+      try (CloseableIterable<FileScanTask> tasks = table.newScan().planFiles()) {
```
Comment (shidayang) [inline flink/v1.12/flink/src/main/java/com/netease/arctic/flink/catalog/ArcticCatalog.java:355]: Mixed table Files is not simple change files + base files. you can use mixed plan

Comment (shidayang) [reply flink/v1.12/flink/src/main/java/com/netease/arctic/flink/catalog/ArcticCatalog.java]: I had talked with the auth of ServerTableDescriptor. The logic is error.

Comment (huyuanfeng2018) [reply flink/v1.12/flink/src/main/java/com/netease/arctic/flink/catalog/ArcticCatalog.java]: > I had talked with the auth of ServerTableDescriptor. The logic is error.

Where can I get some reference about this logic?

Comment (shidayang) [reply flink/v1.12/flink/src/main/java/com/netease/arctic/flink/catalog/ArcticCatalog.java]: You can use KeyedTable#newScan to plan files

Comment (huyuanfeng2018) [reply flink/v1.12/flink/src/main/java/com/netease/arctic/flink/catalog/ArcticCatalog.java]: > You can use KeyedTable#newScan to plan files

It seems cannot get partition or file information through this.

image
There is no `planfiles` method,  and `plantask` cannot get file and partition information.

Comment (shidayang) [reply flink/v1.12/flink/src/main/java/com/netease/arctic/flink/catalog/ArcticCatalog.java]: planTasks -> CombinedScanTask -> KeyedTableScanTask -> ArcticFileScanTask -> PrimaryKeyedFile

Comment (shidayang) [reply flink/v1.12/flink/src/main/java/com/netease/arctic/flink/catalog/ArcticCatalog.java]: planTasks() -> KeyedTableScanTask -> ArcticFileScanTask. ArcticFileScanTask is a FileScanTask.

Comment (huyuanfeng2018) [reply flink/v1.12/flink/src/main/java/com/netease/arctic/flink/catalog/ArcticCatalog.java]: > planTasks() -> KeyedTableScanTask -> ArcticFileScanTask. ArcticFileScanTask is a FileScanTask. or planTasks() -> KeyedTableScanTask -> ArcticFileScanTask -> PrimaryKeyedFile. PrimaryKeyedFile is a ContentFile

done

**Review by shidayang — APPROVED** [review]

**Comment by YesOrNo828** [comment]
@huyuanfeng2018  Thanks for your contribution. 

I find that `ArcticCatalog.listPartitions(ObjectPath tablePath, CatalogPartitionSpec partitionSpec)`  and `ArcticCatalog.listPartitionsByFilter(ObjectPath tablePath, List<Expression> filters)` aren't completed. 

Would you like to open new issues to complete?

**Comment by huyuanfeng2018** [comment]
> @huyuanfeng2018 Thanks for your contribution.
> 
> I find that `ArcticCatalog.listPartitions(ObjectPath tablePath, CatalogPartitionSpec partitionSpec)` and `ArcticCatalog.listPartitionsByFilter(ObjectPath tablePath, List<Expression> filters)` aren't completed.
> 
> Would you like to open new issues to complete?

Okay, I will open a new issue to complete it

**Pull request merged**